README.md
     ____              _          _   
    |  _ \            | |        | |  
    | |_) | ___   ___ | |     ___| |_
    |  _ < / _ \ / _ \| |    / _ \ __|
    | |_) | (_) | (_) | |___|  __/ |_
    |____/ \___/ \___/|______\___|\__|

# Boolaz Log Examination Tool (with anomaly detection)

**The latest version is [BooLET 1.2](https://github.com/boolaz/BooLet/blob/master/booLet_1.2/)** (initially published on 10/10/2016). See [Changelog](https://github.com/boolaz/BooLet/blob/master/CHANGELOG.md) for release notes.

This tool is aimed at optimizing analysis of HTTP logs and finding anomalies in the raw logs in an automated manner, based upon a custom set of Yara rules (SQLi, XSS, directory traversal, shells and PHP shells ...). booLet has the ability to also generate reports and exports based on filters applied to the raw logs.

booLet may be used by forensic investigators, or sysadmins to quickly review HTTP logs and determine the causes of incidents.

[![ScreenShot](https://raw.githubusercontent.com/boolaz/BooLet/master/screenshot/boolet.png)](https://youtu.be/mcDYH6CiiYs)

For a demo video, click on the picture above

-----------------

## Requirements

Boolet 1.2 has been developed in python 2.7 and successfully tested on Linux Ubuntu 14.04 LTS (version 1.1 also works on MacOSX 10.11.6 El Capitan and Windows 8.1x64)

### python modules

Boolet 1.2 requires four additional python modules to work

- [geoip2](https://pypi.python.org/pypi/geoip2) (2.2.0)
- [pyasn](https://pypi.python.org/pypi/pyasn) (1.5.0b6)
- [yara-python](https://pypi.python.org/pypi/yara-python) (3.5.0)
- [PyYAML](https://pypi.python.org/pypi/pyYAML) (3.12)

Installing the required modules can be achieved by the following commands :

    sudo -H pip install geoip2
    sudo -H pip install pyasn
    sudo -H pip install pyYAML
    sudo -H pip install yara-python

### Installing on Windows and Mac

booLet 1.2 is not currently available for those operating systems. A compatible version will be online soon.
For the time being, Windows and mac users may use [booLet 1.1](https://github.com/boolaz/BooLet/tree/master/booLet_1.1) instead but won't benefit from automatic anomaly detection.

-----------------

## How to import your log files

Once you have all modules properly installed, you can import your raw logs.

    $ booLet.py --import combined access.*

For now, booLet supports three formats of logs : ``combined``, ``common``, and ``iponly`` (one IP per line)

The previous command will create a SQLite database, parse your logs, and populate the database with the data. It will also generate a CSV file containing the summary of the imported log files (start, end, nb of lines, nb of unique IP)

During this phase, boolet ignores the following types of file : ``ico jpg png js css gif woff svg robots.txt`` to focus more on static HTML pages and dynamic content such as PHP scripts.

In addition, Boolet ignores visits from most common bots as they are generally not of interest in the scope of a forensic investigation.

    ==> Storing HTTP logs into database <=====================
    file|Start|End|number of lines|nb of unique IP
    www.brunovalentin.com.log|03/Jul/2016 06:37:04|06/Jul/2016 20:07:21|14161|906
    www.brunovalentin.com.log.1|26/Jun/2016 06:53:16|03/Jul/2016 06:36:23|29494|1685

    ==> Generating IP table <=====================

    ==> Generating ASN table <=====================

    ==> Updating geoip and ASN infos <=====================

    ==> Summary <=====================
    Total anomalies in fields : 199/344070

    Anomalies in referer: 7/114694
    Anomalies in uri: 63/114684
    Anomalies in agent: 129/114692

    directory traversal (dirtrav) : 8
    SQL Injection attempt (sqli) : 1
    shell access attempt (shell) : 132
    encoded string (encoded) : 57
    Unusual long field (longfield) : 2

You are now ready to submit your requests to the database.

Each and every line of log is now associated with additional fields regarding IP addresses (country name, country ISO code, city, Autonomous System (ASN), ASN description, and IP range the IP address belongs to)

-----------------

## How to query data

you can retrieve your log data based upon selected fields in the requested order

example

    $ booLet.py --fields dhicu

will give you the following result

    ...
    2016-07-06|19:58:29|194.187.170.111|FR|/page/2/
    2016-07-06|19:58:32|194.187.170.111|FR|/a-propos/
    2016-07-06|19:58:37|194.187.170.111|FR|/dev/boofadet/
    2016-07-06|19:59:49|208.78.55.117|MF|/mac/clavier-mac-francais-windows/
    2016-07-06|19:59:55|194.187.170.111|FR|/category/trucs-de-geek/
    ...

The following fields may be used :

    i: ip
    d: date
    h: time
    m: method
    u: url
    t: http status code
    z: size
    r: referer
    a: agent
    c: country code
    n: country name
    y: city
    s: asn
    g: asn range
    l: asn label
    o: type of anomaly detected

-----------------

## How to filter output

If you want to narrow down your search and to be presented with more accurate information, you can also filter data upon specific fields

This example displays fields ``dhiuns`` where ``country code=(BE or ES or DE)`` and ``time begins with "16:"``

    $ booLet12.py --fields dhiuns --country BE,ES,DE --time 16:

output :

    2016-06-28|16:27:10|136.243.56.239|/feed/|Germany|AS24940
    2016-06-28|16:27:10|136.243.56.239|/feed/|Germany|AS24940
    2016-06-29|16:08:24|85.26.90.2|/|Belgium|AS12392
    2016-06-29|16:10:11|85.26.90.2|/investigation-numerique/encase-forensic-8-fin-juin/|Belgium|AS12392
    2016-06-29|16:10:13|85.26.90.2|/investigation-numerique/effacement-chimique-disque-dur/|Belgium|AS12392
    2016-06-29|16:11:21|85.26.90.2|/page/2/|Belgium|AS12392
    2016-06-29|16:11:22|85.26.90.2|/ul/monitorssetup-1024x669.jpeg|Belgium|AS12392
    2016-06-29|16:13:18|85.26.90.2|/sans-categorie/des-machines-virtuelles-oui-mes-des-ecrans-cest-mieux/|Belgium|AS12392
    2016-06-29|16:13:19|85.26.90.2|/securite-info/rechercher-traces-meterpreter-en-memoire-volatility/|Belgium|AS12392
    2016-06-29|16:24:54|136.243.56.239|/feed/|Germany|AS24940
    2016-06-30|16:27:31|136.243.56.239|/feed/|Germany|AS24940
    2016-07-06|16:30:29|109.128.211.78|/|Belgium|AS5432

Of course you're free to chose the fields order : ``dhiuns`` is not ``idhsnu``

You can also search for lines with anomalies

    Ex: boolet12.py -f oiur --anomaly sqli,dirtrav
    Ex: boolet12.py -f oiur --anomaly all

### example with anomalies

The following example will retrieve only SQL Injections,SQL file downloads and directory traversals and will display the following fields (odhiu : anomaly, date, time, ip, uri)

    booLet12.py -f odhiu --anomaly sqlfile,sqli,dirtrav

The result will be as following :

    sqlfile|2014-12-07|04:24:49|46.161.41.257|/Agenda.sql
    encoded,shell,sqli|2014-12-10|17:21:19|61.182.202.257|/insert+into%3A%2C%45%56%76%54%4E%3A%2E%45
    dirtrav|2014-12-03|10:10:22|74.208.69.198|/vtigercrm/modules/com_vtiger_workflow/sortfieldsjson.php?module_name=../../../../../../../..//etc/asterisk/sip_additional.conf%00
    dirtrav|2014-12-18|21:34:49|202.62.155.257|/cart.php?a=byroe&templatefile=../../../configuration.php%00
    dirtrav|2014-12-23|20:43:49|192.99.68.257|/?lang=../../../../../../../../../../../proc/self/environ
    dirtrav|2014-12-23|20:46:04|192.99.68.257|/?lang=../../../../../../../../../../../proc/self/environ%00
    dirtrav|2015-01-04|23:56:34|82.97.16.257|//cgi-bin/webcm?getpage=../html/menus/menu2.html&var:lang=%26%20allcfgconv%20-C%20voip%20-c%20-o%20-%20../../../../../var/tmp/voip.cfg%20%26
    dirtrav|2015-01-05|02:19:43|82.97.16.257|//cgi-bin/webcm?getpage=../html/menus/menu2.html&var:lang=%26%20allcfgconv%20-C%20voip%20-c%20-o%20-%20../../../../../var/tmp/voip.cfg%20%26
    dirtrav|2015-01-05|02:55:11|82.97.16.257|//cgi-bin/webcm?getpage=../html/menus/menu2.html&var:lang=%26%20allcfgconv%20-C%20voip%20-c%20-o%20-%20../../../../../var/tmp/voip.cfg%20%26
    dirtrav|2015-01-07|00:10:00|216.121.118.257|/index.php?page=../../../../../../../../../../../proc/self/environ

use the --help option of boolet for more options

-----------------

## How to export output data

When you're happy with your output, you have two options to save it to an output file.

- Use the > sign to redirect to a new file

    ``$ booLet.py --fields dhi > outfile.csv``

- Boolet has also a ``--out`` option to export in a csv file (pipe separated fields)

    ``$ booLet.py --fields dhi --out outfile.csv``

Both will generate the following file, named outfile.csv :

    2016-06-26|06:54:10|146.185.251.48
    2016-06-26|06:57:39|146.185.251.210
    2016-06-26|07:00:09|94.228.34.250
    2016-06-26|07:00:21|146.185.251.48
    2016-06-26|07:00:35|37.187.109.125
    2016-06-26|07:00:36|91.134.167.121
    2016-06-26|07:01:04|86.247.45.90
    ...

This format has been adopted because it is as easy to read as to process with bash tools like awk, cut or sort.

## Todo list

- update functionality for geoip and asn databases
- add more input formats

Stay tuned for updates and please, feel free to report any bug to the author.


CHANGELOG.md
CHANGELOG
===========================

* v1.2  adding yara rules support, adding automatic anomaly detection, adding yaml configuration file

* v1.1  parsing pattern for combined log has been improved, a windows binary has been generated.

* v1.0  initial version of booLET


booLet_1.2/src/BooLetLib.py
# !/usr/bin/python
# -*- coding: utf-8 -*-
import re,sys,os
import yara

# -----------------------------------------------------------
usage_text="""IMPORTING RAW LOGS AND CREATING LOG DATABASE     
booLet.py --import %FORMAT_TYPE% logfile_name     

FORMAT TYPES 
  combined : apache combined
  common : apache common
  iponly : list of IP addresses
       
EXTRACTING COLUMNS     
 booLet.py --field idhmutzracnysgl    
  i: ip     
  d: date     
  h: time     
  m: method     
  u: url     
  t: http status code     
  z: size     
  r: referer     
  a: agent     
  c: country code     
  n: country name     
  y: city     
  s: asn     
  g: asn range     
  l: asn label     
  o: type of anomaly
       
CONDITIONS AND FILTERS     
 booLet.py --ip 41.123.12.43,132,23,56,131 --agent msie --anomaly sqli,dirtrav    
  --ip @IP    
  --date XXXX-XX-XX    
  --time XX:XX:XX     
  --method     
  --uri     
  --status     
  --referer     
  --agent     
  --asn     
  --country ISO_country_code     
  --isp     
  --anomaly XX     
        all : display all anomalies 
        sqli : SQL injections
        ... : any type of anomaly detected during the scans
       
EXPORTING RESULTS     
 booLet.py --out output.csv
"""

# -----------------------------------------------------------
def usage(usage_text):
    """Usage"""
    print usage_text

# --------------------------------------
def title(title_label):
  """display title"""
  print """
==> {} <=====================""".format(title_label)

# -----------------------------------------------------------
class Banner(object):
    """Banner for the program"""
    def __init__(self, banner_values):
        super(Banner, self).__init__()
        self.banner_values = banner_values

    def display(self):
        print("""
****************************************************** \n\
* {0} {1} ({2}) \n\
* {3}           \n\
* Author : Bruno Valentin (bruno@boolaz.com) \n\
* Updates : {4} \n\
******************************************************""" \
         .format(self.banner_values['name'],self.banner_values['version'],
                 self.banner_values['release'],self.banner_values['purpose'],
                 self.banner_values['link']))

# --------------------------------------
class Anomalies(object):
    """This class deals with anomalies in logs"""
    anomaly_line_nb=0
    anomalies_nb=0
    processed_nb=0
    anomaly_nb_by_type=dict()
    description=dict()

    def __init__(self, log_line):
        """class init"""
        super(Anomalies, self).__init__()

    def display(self):
        """display None"""
        return False

    @staticmethod
    def stats():
        """stats on anomalies"""

        # displays global stats
        print ("Total anomalies in fields : {}/{}"
          .format(Anomalies.anomaly_line_nb,Anomalies.get_nb_processed()))
        print

        # displays values for each child class of Anomaly
        for child_object in [cls for cls in Anomalies.__subclasses__()]:
            child_object.stats_by_field()
        print

        # then displays values for each anomaly identified
        for anomaly,value in Anomalies.anomaly_nb_by_type.items():
            print("{} ({}) : {}".format(Anomalies.description[anomaly], 
                anomaly, value))
        return True

    @classmethod
    def compile_ruleset(cls):
        """compile the rules from the yara files"""
        script_path=os.path.dirname(os.path.realpath(__file__))
        anomalies_file=script_path+"/"+cls.rules_filename
        cls.anomalies_rules = yara.compile(anomalies_file,
          externals= {
          'strval': '', 
          'strlen': 0
          })
        return True

    @classmethod
    def stats_by_field(cls):
        """generates stats for each field"""
        print ("Anomalies in {}: {}/{}"
            .format(cls.field_label,cls.get_nb_anomalies()
              ,cls.get_nb_processed()))

    @classmethod
    def count_anomalies(cls,value):
        """count anomalies in the current sub-class and the parent class"""
        if value==0:
            cls.anomalies_nb=1
            Anomalies.anomalies_nb=1
        else:
            cls.anomalies_nb+=value
            Anomalies.anomalies_nb+=value

    @classmethod
    def get_nb_anomalies(cls):
        """ return nb of anomalies in current class"""
        return "{}".format(cls.anomalies_nb)

    @classmethod
    def get_nb_processed(cls):
        """ return nb of processed lines in current class"""
        return "{}".format(cls.processed_nb)

    def search(self):
        """matches the line against the anomalies ruleset"""
        matches = self.anomalies_rules.match(data=self.log_line,
          externals= {
          'strval': self.log_line, 
          'strlen': len(self.log_line)
          })
        try:
            yaras=dict()
            yaras=matches['main']
        except:
            matching_causes=list()
        else:
            Anomalies.anomaly_line_nb+=1
            self.count_anomalies(1)
            matching_causes=list()
            for yara in yaras:
                matching_causes.append(yara['rule'])
                description=yara['meta']['description']
                if yara['rule'] not in Anomalies.anomaly_nb_by_type:
                    Anomalies.description[yara['rule']]=description
                    Anomalies.anomaly_nb_by_type[yara['rule']]=1
                else:
                    Anomalies.anomaly_nb_by_type[yara['rule']]+=1

        return(matching_causes)

# --------------------------------------
class AnomRef(Anomalies):
    """This class deals with anomalies in referer"""

    # variables which are specific to this subclass
    field_label="referer"
    rules_filename="anomaly_referer.yar"
    anomalies_nb=0
    processed_nb=0

    def __init__(self, log_line):
        """counts processed lines (current + parent)"""
        self.log_line = log_line
        self.__class__.processed_nb+=1
        self.__class__.__base__.processed_nb+=1

# --------------------------------------
class AnomUri(Anomalies):
    """This class deals with anomalies in uri"""

    # variables specific to this subclass
    field_label="uri"
    rules_filename="anomaly_uri.yar"
    anomalies_nb=0
    processed_nb=0

    def __init__(self, log_line):
        """counts processed lines (current + parent)"""
        self.log_line = log_line
        self.__class__.processed_nb+=1
        self.__class__.__base__.processed_nb+=1

# --------------------------------------
class AnomAgt(Anomalies):
    """This class deals with anomalies in Agent field"""

    # variables specific to this subclass
    field_label="agent"
    rules_filename="anomaly_agent.yar"
    anomalies_nb=0
    processed_nb=0

    def __init__(self, log_line):
        """counts processed lines (current + parent)"""
        self.log_line = log_line
        self.__class__.processed_nb+=1
        self.__class__.__base__.processed_nb+=1


booLet_1.2/booLet12.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# ---------------------------------------------------------
# Name       : BooLET.py
# Purpose    : Log Examination Tool with anomaly detection
# Revision   : 1.2 (01/09/2016)
# Author     : Bruno Valentin - bruno@boolaz.com
# Updates    : https://github.com/boolaz/BooLet
# Reference  : http://www.brunovalentin.com/dev/
# ---------------------------------------------------------

import os,sys,re,getopt
from sys import argv
from datetime import datetime
import geoip2.database
import sqlite3 as lite
import pyasn, yaml
from src.booLetLib import *

reload(sys)
sys.setdefaultencoding("utf-8")

softdesc={"name":"booLet",
          "version":"1.2", \
          "release":"01/09/2016", \
          "purpose":"Log Examination Tool with automatic anomaly detection", \
          "link":"https://github.com/boolaz/booLet" }

log_formats = {
    'common':'([^ ]+) [^ ]+ [^ ]+ \[([^:]+):([^ ]+) (.*?)\] "([^ ]+) ?([^ ]*) ?([^ ]*)" (\d+) ([\d\-]+)(.*?)(.*?)',
    'combined':'^(\S+) \S+ \S+ \[([^:]+):(\d+:\d+:\d+) ([^\]]+)\] "(\S+) ?(\S+)? ?(\S+).*" (\S+) (\S+) "(.+)?" "(.*)"$',
    'iponly':'([^|\n]+)(.*?)(.*?)(.*?)(.*?)(.*?)(.*?)(.*?)(.*?)(.*?)(.*?)'
}

# -----------------------------------------------------------
def create_new_database(db):  # creation d'une base vide
    with db:
        db.execute("DROP TABLE IF EXISTS logs")
        db.execute("CREATE TABLE logs(IP TEXT, ladate TEXT,lheure TEXT, \
         method TEXT,uri TEXT, status INT, size INT, referer TEXT, agent TEXT, \
         anomflags TEXT)")
        db.execute("CREATE INDEX _IP_LOGS ON logs(IP ASC)")
        db.execute("CREATE INDEX _ladate_LOGS ON logs(ladate ASC)")
        db.execute("CREATE INDEX _lheure_LOGS ON logs(lheure ASC)")
        db.execute("CREATE INDEX _method_LOGS ON logs(method ASC)")
        db.execute("CREATE INDEX _uri_LOGS ON logs(uri ASC)")
        db.execute("CREATE INDEX _status_LOGS ON logs(status ASC)")
        db.execute("CREATE INDEX _flags_LOGS ON logs(anomflags)")

# -----------------------------------------------------------
def format_line(line,log_format):

    global log_formats

    regex=log_formats[log_format]
    try:
        logip,logdate,logtime,logtz,logmethod,loguri,logversion,logstatus, \
        logsize,referer,logagent=re.match(regex, line).groups()
        # print re.match(regex, line).groups()
    except:
        print u"The following line raises an exception : \n{0}".format(line)
        # print re.match(regex, line).groups()
        sys.exit(2)
    return(logip,logdate,logtime,logtz,logmethod,loguri,logversion,logstatus,
        logsize,referer,logagent)

# -----------------------------------------------------------
def upload_logs_in_db(db,outdir,log_format,fichiers):

    excluded_agents=".*({}).*".format("|".join(boocfg['crawl_excl']['list']))
    excluded_uri=".*\.({})(\?|&)?".format("|".join(boocfg['uri_excl']['list']))
    uri_comp = re.compile(excluded_uri, flags=re.I)
    agents_comp = re.compile(excluded_agents, flags=re.I)

    # global log_formats
    db.row_factory = lite.Row  # @UndefinedVariable

    title("Storing HTTP logs into database")

    # summary.csv 
    resume_file = open(outdir + '/' + 'summary.csv',"w")
    resume_line=u"{0}|{1}|{2}|{3}|{4}".format("file","Start","End",
        "number of lines","nb of unique IP")
    print resume_line
    resume_file.write("SUMMARY\n")
    resume_file.write(resume_line+"\n")

    # for each logfile --> processing lines of log
    detail_fichiers=[]
    for fichier in fichiers:

        fd = open(fichier,"r")
        lignes_total= 0
        for line in fd:
            lignes_total+= 1

        # compiles rulesets
        AnomUri.compile_ruleset()
        AnomRef.compile_ruleset()
        AnomAgt.compile_ruleset()

        nb_lines=0
        ip=[]
        les_lignes=[]
        infile = open(fichier,"r")
        for line in infile:
            nb_lines+=1

            # matches the line
            # print u"{0} : {1}/{2}".format(fichier,nb_lines,lignes_total)
            logip,logdate,logtime,logtz,logmethod,loguri,logversion,logstatus,\
             logsize,referer,logagent=format_line(line,log_format)

            # determines beginning and end of the time period
            if nb_lines==1: debut=u"{0} {1}".format(logdate,logtime)
            fin=u"{0} {1}".format(logdate,logtime)
            if logdate:
                logdate = u"{0}".format(datetime.strptime(logdate, "%d/%b/%Y"))
                date_stamp = logdate[0:logdate.find(' ')]
            else:
                date_stamp=''

            # unknown ip --> appends ip array
            if logip not in ip: ip.append(logip)

            # Discards the line or not ?
            uri_discard=False
            if boocfg['uri_excl']['discard']:
                if uri_comp.match("{0!s}".format(loguri)):
                    uri_discard=True
            crawl_discard=False
            if boocfg['crawl_excl']['discard']:
                if agents_comp.match("{0!s}".format(logagent)):
                    crawl_discard=True

            # if no reason to discard the line
            if not (uri_discard or crawl_discard):
                
                # matches against yara rules
                if (loguri):
                    MyAnomUri=AnomUri("{}".format(loguri))
                    my_anom_uri_flags=MyAnomUri.search()

                if (referer):
                    MyAnomRef=AnomRef("{}".format(referer))
                    my_anom_ref_flags=MyAnomRef.search()

                if (logagent):
                    MyAnomAgt=AnomAgt("{}".format(logagent))
                    my_anom_agt_flags=MyAnomAgt.search()

                # concat lists without duplicates
                my_flags=list(set().union(my_anom_uri_flags,my_anom_agt_flags,
                    my_anom_ref_flags))

                # converts lists to comma separated strings
                if (my_flags):
                    for i, item in enumerate(my_flags):
                        my_flags[i] = "-{}-".format(item)
                    insert_flags=','.join(my_flags)
                else :
                    insert_flags=''

                # adds to les_lignes
                les_lignes.append((logip,date_stamp,logtime,logmethod, \
                    loguri,logstatus,logsize, referer, logagent,insert_flags))


        # inserts lines
        with db:
            cur = db.cursor()
            sql="INSERT INTO logs VALUES (?,?,?,?,?,?,?,?,?,?);"
            db.executemany(sql,les_lignes)

        # Summary --> file summary.csv
        detail_fichiers.append((os.path.basename(fichier),debut,fin,nb_lines,
            len(ip)))
        resume_line=u"{0!s}|{1!s}|{2!s}|{3!s}|{4!s}".format(
            os.path.basename(fichier),debut,fin,nb_lines,len(ip))
        print resume_line
        resume_file.write(resume_line+"\n")
        infile.closed

    resume_file.closed


# -----------------------------------------------------------
def create_ip_table(db):

    title("Generating IP table")

    ip_list=[]

    with db:
        cur = db.cursor()

        db.execute("DROP TABLE IF EXISTS ips")
        db.execute("""CREATE TABLE ips(IP TEXT, isocode TEXT, country_name TEXT,
                      city_name TEXT, asn INT, range TEXT, asnlabel TEXT)""")
        db.execute("CREATE UNIQUE INDEX _IP_IPS ON ips(IP ASC)")
        db.execute("CREATE INDEX _isocode_IPS ON ips(isocode ASC)")
        db.execute("CREATE INDEX _asn_IPS ON ips(asn ASC)")

        sql="SELECT DISTINCT IP FROM logs;"
        results=cur.execute(sql)
        for mon_ip in results:
            ip_list.append((mon_ip[0],'-','-','-','-','-','-'))
        sql="INSERT INTO ips VALUES (?,?,?,?,?,?,?);"
        db.executemany(sql,ip_list)


# -----------------------------------------------------------
def create_asn_table(db):

    title("Generating ASN table")

    fichier=script_path+"/booLet_extres/ASNList.txt"
    asn_list=[]
    les_lignes=[]

    infile = open(fichier,"r")
    for line in infile:
        regex = '([^ ]+) +(.+)'
        asn,asnLabel=re.match(regex, line).groups()
        asnlabel="%s" % unicode(asnLabel)  # @UndefinedVariable
        les_lignes.append((asn,asnlabel))

    with db:
        db.execute("DROP TABLE IF EXISTS asns")
        db.execute("CREATE TABLE asns(asn TEXT, asnlabel TEXT)")
        db.execute("CREATE UNIQUE INDEX _asn_asns ON asns(asn ASC)")

        sql="INSERT INTO asns VALUES (?,?);"
        db.executemany(sql,les_lignes)

    infile.closed


# -----------------------------------------------------------
def maj_location_asn(db):

    title("Updating geoip and ASN infos")

    reader = geoip2.database.Reader(script_path+
        '/booLet_extres/GeoLite2-City.mmdb')
    asn_db = pyasn.pyasn(script_path+'/booLet_extres/ipasn_20150224.dat')
    les_lignes=[]
    with db:
        cur = db.cursor()
        sql="SELECT IP FROM ips;"
        results=cur.execute(sql)
    for mon_ip in results:
        country_name=city_name=iso_code='-'

        try:
            response = reader.city(mon_ip[0])
        except:
            country_name=city_name=iso_code='-'
        else:
            if response.country.names:
                if "fr" in response.country.names: 
                    country_name=response.country.names['fr']
                if response.country.names!='' : 
                    country_name=response.country.name

            if response.city.name:
                if response.city.name!='':
                    city_name=response.city.name
                else: city_name='-'

            if response.country.iso_code in ['','None'] : iso_code='-'
            else: iso_code="{0!s}".format(response.country.iso_code)

        try:
            asn,range=asn_db.lookup(mon_ip[0])
        except:
            asn=range='-'
        else:
            if ("{0!s}".format(asn)).lower()=='none': asn=range='-'
            else: asn="AS{0!s}".format(asn)

        les_lignes.append((iso_code,country_name,city_name,asn,range,mon_ip[0]))

    with db:
        cur = db.cursor()
        sql="""UPDATE ips SET isocode=?, country_name=?, city_name=?, asn=?, 
               range=? WHERE IP=?"""
        db.executemany(sql,les_lignes)

        sql="""SELECT DISTINCT(i.asn),a.asnlabel FROM ips i, asns a 
               WHERE i.asn = a.asn;"""
        results=cur.execute(sql)
        values=[]
        for result in results:
            values.append((result[1],result[0]))
        sql="UPDATE ips SET asnlabel=? WHERE asn=?"
        db.executemany(sql,values)

# -----------------------------------------------------------
def get(db,fields='idhmutzracnysglo',condition='',outfile=''):

    if outfile:
        dest = open(outfile,"w")
        print ("/--------------------------------/ \n\
/     Exporting output file      /\n\
/--------------------------------/")
    if condition: condition="WHERE "+condition

    if fields=='all': fields='idhmutzracnysgl'
    fields_nb=len(fields)

    db.row_factory = lite.Row  # @UndefinedVariable

    with db:
        cur = db.cursor()
        sql=("SELECT logs.*,  ips.* FROM logs LEFT JOIN ips ON ips.IP=logs.IP "
            +condition+" ORDER BY logs.ladate ASC, logs.lheure ASC;")
        lignes=cur.execute(sql)

    if outfile:
        header_line=""
        cpt=0
        for field in fields:
            cpt+=1
            if   field in "i" : header_line+="IP Address"
            elif field in "d" : header_line+="Date"
            elif field in "h" : header_line+="Time"
            elif field in "m" : header_line+="Method"
            elif field in "u" : header_line+="URI"
            elif field in "t" : header_line+="Status code"
            elif field in "z" : header_line+="Size"
            elif field in "r" : header_line+="Referer"
            elif field in "a" : header_line+="Agent"
            elif field in "c" : header_line+="Country code"
            elif field in "n" : header_line+="Country"
            elif field in "y" : header_line+="City"
            elif field in "s" : header_line+="ASN"
            elif field in "g" : header_line+="IP range"
            elif field in "l" : header_line+="Operator name"
            elif field in "o" : header_line+="Anomalies"
            if cpt<fields_nb: header_line+="|"
        dest.write(header_line+"\n")

    for ligne in lignes:
        (isocode,country_name,city_name,asn,range,asnlabel)=(
            "%s" % ligne["isocode"],"%s" % ligne["country_name"],
            "%s" % ligne["city_name"],"%s" % ligne["asn"],"%s" % ligne["range"],
            "%s" % ligne["asnlabel"])

        output_line=""
        cpt=0
        for field in fields:
            cpt+=1
            if   field in "i" : output_line+="%s" % ligne["IP"]
            elif field in "d" : output_line+="%s" % ligne["ladate"]
            elif field in "h" : output_line+="%s" % ligne["lheure"]
            elif field in "m" : output_line+="%s" % ligne["method"]
            elif field in "u" : output_line+="%s" % ligne["uri"]
            elif field in "t" : output_line+="%s" % ligne["status"]
            elif field in "z" : output_line+="%s" % ligne["size"]
            elif field in "r" : output_line+=ligne["referer"]
            elif field in "a" : output_line+=ligne["agent"]
            elif field in "c" : output_line+=isocode
            elif field in "n" : output_line+=country_name
            elif field in "y" : output_line+=city_name
            elif field in "s" : output_line+=asn
            elif field in "g" : output_line+=range
            elif field in "l" : output_line+=asnlabel
            elif field in "o" : 
                # gets rid of - characters around each flag
                anomflags=re.sub(r'^-|-$', '', ligne["anomflags"])
                anomflags=re.sub(r'-,-', ',', anomflags)
                output_line+="%s" % anomflags
            if cpt<fields_nb: output_line+="|"

        output_line=output_line.encode('UTF-8')
        if outfile: dest.write(output_line+"\n")
        else: print output_line

# -----------------------------------------------------------
def make_clause(field,items):
    clause="("
    list=("%s" % items).split(",")
    cpt=0
    for item in list:
        cpt+=1
        if field in ("logs.ladate", "logs.lheure","logs.IP"): 
            clause+="%s LIKE '%s%%'" % (field,item)
        elif field in ("logs.uri","logs.agent","logs.referer","ips.asnlabel"): 
            clause+="%s LIKE '%%%s%%'" % (field,item)
        elif field in ("logs.anomflags"):
            if (item=='all'):
                clause+="%s is not ''" % (field)
            else:
                # adds "-" in front of, and behind item to match items in DB
                item="-{}-".format(item)
                clause+="%s LIKE '%%%s%%'" % (field,item)
        else : 
            clause+="%s ='%s'" % (field,item)
        if cpt < len(list): clause+=" OR "
    clause+=")"
    return(clause)

# -----------------------------------------------------------
def ready():
    print "\nDatabase is now ready for querying..."
    print "Use boolet.py -f ??? to query"
    print "Ex: boolet.py -f dhicl --country FR,ES"
    print "Ex: boolet.py -f oiur --anomaly sqli,dirtrav\n"
    return True

# -----------------------------------------------------------
def main(argv):
    #global db_file
    global log_formats

    chemin,fields,importation,columns,out=('','','','','')
    clauses=[]

    # Banner
    my_banner=Banner(softdesc)
    my_banner.display()

    # arguments and options
    try:
        opts, args = getopt.getopt(argv, \
            "hf:i:o:", \
            ["help","fields=","import=","out=","ip=","date=","time=","method=",
            "uri=","status=","referer=","agent=","asn=","country=","isp=",
            "anomaly="])
    except getopt.GetoptError:
        usage(usage_text)
        sys.exit(2)

    if opts :
        for opt, arg in opts:
            if opt in ("-h", "--help"):
                usage(usage_text)
                sys.exit()
            elif opt in ("-i", "--import"):
                log_format = "%s" % arg
                if log_formats.has_key(log_format):
                    importation=1
                else:
                    print ("Format type \"{0}\" doesn't exist. BooLET is using combined format instead").format(log_format)
                    log_format = "combined"
                    importation=1
            elif opt in ("-f", "--fields"):
                columns = "%s" % arg
            elif opt in ("-o", "--out"):
                out= "%s" % arg

            elif opt in ("--ip"):
                clauses.append(make_clause('logs.IP',"%s" % arg))
            elif opt in ("--date"):
                clauses.append(make_clause('logs.ladate',"%s" % arg))
            elif opt in ("--time"):
                clauses.append(make_clause('logs.lheure',"%s" % arg))
            elif opt in ("--method"):
                clauses.append(make_clause('logs.method',"%s" % arg))
            elif opt in ("--uri"):
                clauses.append(make_clause('logs.uri',"%s" % arg))
            elif opt in ("--status"):
                clauses.append(make_clause('logs.status',"%s" % arg))
            elif opt in ("--referer"):
                clauses.append(make_clause('logs.referer',"%s" % arg))
            elif opt in ("--agent"):
                clauses.append(make_clause('logs.agent',"%s" % arg))
            elif opt in ("--asn"):
                clauses.append(make_clause('ips.asn',"%s" % arg))
            elif opt in ("--country"):
                clauses.append(make_clause('ips.isocode',"%s" % arg.upper()))
            elif opt in ("--isp"):
                clauses.append(make_clause('ips.asnlabel',"%s" % arg))
            elif opt in ("--anomaly"):
                clauses.append(make_clause('logs.anomflags',"%s" % arg))

            globalclause=" AND ".join(clauses)
    else:
        usage(usage_text)
        sys.exit(2)

    rep_travail=os.path.dirname(chemin)
    if rep_travail=="": rep_travail="."

    DB = lite.connect(rep_travail + "/" + boocfg['files']['db_file'])  

    # importation of data
    if importation:
        create_new_database(DB)
        upload_logs_in_db(DB,rep_travail,log_format,args)
        create_ip_table(DB)
        create_asn_table(DB)
        maj_location_asn(DB)

        title("Summary")
        Anomalies.stats()
        ready()

    # get by columns
    if columns:
        get(DB,fields=columns,outfile=out,condition=globalclause)

if __name__ == "__main__":
    script_path=os.path.dirname(os.path.realpath(__file__))
    with open(script_path+"/src/config.yml", 'r') as ymlfile:
      boocfg=yaml.load(ymlfile)
    main(sys.argv[1:])
